### PR TITLE
Reorg crosschain module

### PR DESCRIPTION
**Pull request by shiwk**

### CODE REVIEW

**Comment by shiwk** [comment]
image

File: AElf.Types/Protobuf/Proto/kernel.proto [hunk 1]
```diff
@@ -168,14 +172,15 @@ message BlockHeader {
     bytes P = 8;
     google.protobuf.Timestamp Time = 9;
     int32 ChainId = 10;
-    Hash SideChainTransactionsRoot = 11;
+    BlockExtraData BlockExtraData = 11;
```
Comment (shiwk) [inline AElf.Types/Protobuf/Proto/kernel.proto]: Add _BlockExtraData_ here.

File: AElf.Contracts.CrossChain/CrossChainContract_Helper.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+using AElf.Sdk.CSharp.State;
+using AElf.Types.CSharp;
+using Google.Protobuf;
+using Google.Protobuf.WellKnownTypes;
+
+namespace AElf.Contracts.CrossChain
+{
+    public partial class CrossChainContract
+    {
+        private Hash Propose(string proposalName, double waitingPeriod, Address fromAddress,
```
Comment (loning) [inline AElf.Contracts.CrossChain/CrossChainContract_Helper.cs]: don't use double

File: AElf.CrossChain.Core/Helpers/ICrossChainContractReader.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        }
+
+        public Task<ulong> GetBoundParentChainHeightAsync(ulong localChainHeight)
+        {
+            throw new System.NotImplementedException();
+        }
+
+        public async Task<ulong> GetParentChainCurrentHeightAsync(int chainId, Hash previousBlockHash, ulong preBlockHeight)
+        {
+            var crossChainContractMethodAddress = ContractHelpers.GetCrossChainContractAddress(chainId);
```
Comment (loning) [inline AElf.CrossChain.Core/Helpers/ICrossChainContractReader.cs]: ContractHelpers should be removed.

cross chain address should get by service and saved/cached in your contract state

File: AElf.Contracts.Authorization/IMessageExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
+using System;
+using System.Collections.Generic;
+using Google.Protobuf;
+
+namespace AElf.Contracts.Authorization
+{
+    public static class IMessageExtensions
```
Comment (loning) [inline AElf.Contracts.Authorization/IMessageExtensions.cs]: remove this extension in the future, you should manage null & empty.

it's better not store empty value.

**Pull request merged**